### PR TITLE
Additional Error Codes, Messages, Descriptions and Solutions

### PR DESCRIPTION
**Pull request by Spletz47**
Added all 160 and 165 error codes and also added missing messages for existing 160/165 error codes.

### CODE REVIEW

**Comment by jonbarrow** [comment]
This looks good! Out of curiosity, where did these messages come from? The existing messages were generated by parsing the MSBT files from the Wii U and 3DS error modules, so I'm curious where these ones were at

**Comment by Spletz47** [comment]
I do apologise for not following the contribution guide correctly. There are more error codes that need to be filled out, which I was planning on doing later on, so I'll make sure to correctly follow it from now into the future

160 error codes are actually listed in error.rpx, I assume so that if there is a fatal system error, it doesn't need to look further through the files. For the purposes of this, since I'm not smart enough to figure out which message was related to which error code, I just scoured the internet and used photos of people's errors as reference.

**Comment by jonbarrow** [comment]
Sounds good. I've also changed the target branch to `dev`, like it's supposed to be. Please ensure that things are still good

File: data/102/1021/en_US.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
 			"message": "Use of this feature requires a linked\nNintendo Network ID.\n\nYou can link an ID from User Settings.",
 			"short_description": "Unknown cause.",
 			"long_description": "Unknown cause.",
-			"short_solution": "Unknown solution.",
+			"short_solution": "Link a Nintendo/Pretendo Network ID to your user.",
```
Comment (jonbarrow) [inline data/102/1021/en_US.json:8]: This goes for all the errors which you have done this with. If you are going to add a short solution, we ask that you also add a long solution. The long solution is what is going to be shown on the website, and supports markdown rendering

File: data/102/0001/en_US.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
 		"0001": {
 			"name": "MAIL_ADDRESS_NOT_CONFIRMED",
 			"message": "",
-			"short_description": "Unknown cause.",
+			"short_description": "Unknown cause",
```
Comment (jonbarrow) [inline data/102/0001/en_US.json:6]: Why?

File: data/102/2595/en_US.json [hunk 3]
```diff
@@ -3,9 +3,9 @@
 		"2595": {
 			"name": "MII_NOT_REGISTERED",
 			"message": "You have not registered a Mii.",
-			"short_description": "Unknown cause.",
+			"short_description": "A Mii has not been registred.",
```
Comment (jonbarrow) [inline data/102/2595/en_US.json:6]: Might want to add that it's a Mii registered to an account which is missing

File: data/102/2612/en_US.json [hunk 4]
```diff
@@ -3,9 +3,9 @@
 		"2612": {
 			"name": "WRONG_MAIL_ADDRESS",
 			"message": "The e-mail address you entered is not\ncorrect.\n\nPlease check the e-mail address\nyou entered and try again.",
-			"short_description": "Unknown cause.",
+			"short_description": "The entered e-mail address is incorrect.",
```
Comment (jonbarrow) [inline data/102/2612/en_US.json:6]: Just like with the short solutions, if you're going to add a short description then we also ask that you add a long description. This, too, supports markdown and is intended for the website

File: data/102/2613/en_US.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
 			"message": "The information entered is not correct.\nYou may have entered the ID of another\nservice.\n\nPlease enter the details again, making\nsure that the Nintendo Network ID,\npassword, and e-mail address are\ncorrect.",
 			"short_description": "Invalid account information provided. Either your username, password, or email address does not match",
 			"long_description": "Unknown cause.",
-			"short_solution": "Enter corrected information and try again",
+			"short_solution": "Enter corrected information and try again. You might also not be connected to Pretendo Network.",
```
Comment (jonbarrow) [inline data/102/2613/en_US.json:8]: I would change this to be a bit more generic. If someone is on NN and tries to use their PNID, or connected to Pretendo and tries to use their NNID, they get the same error here. But this makes it sound like it only happens when not using Pretendo

File: data/102/2882/en_US.json [hunk 6]
```diff
@@ -5,7 +5,7 @@
 			"message": "This online service is no longer\navailable.\n\nSorry for the inconvenience.",
 			"short_description": "The server you requested is closed and no longer maintained",
 			"long_description": "Unknown cause.",
-			"short_solution": "Self host the server or use another hosted instance",
+			"short_solution": "Self host the server or use another hosted instance. You also may not be connected to Pretendo Network.",
```
Comment (jonbarrow) [inline data/102/2882/en_US.json:8]: See my previous comment about making this more generic

File: data/199/0106/en_US.json [hunk 7]
```diff
@@ -3,9 +3,9 @@
 		"0106": {
 			"name": "Unknown",
 			"message": "You have not entered anything.",
-			"short_description": "Unknown cause.",
+			"short_description": "You have failed to enter what has been requested.",
```
Comment (jonbarrow) [inline data/199/0106/en_US.json:6]: This seems like a bit of an odd way to phrase "you used an empty input"

**Review by jonbarrow — CHANGES_REQUESTED** [review]
Overall seems good, just a few requests. Also how many more errors do you plan to add to this PR? Just so I can know when to actually do a final review and merge

File: data/102/0001/en_US.json (hunk 2 above)
Comment (Spletz47) [reply data/102/0001/en_US.json:6]: Entered something, realised it was wrong, re-entered placeholder, forgot the full stop

**Comment by Spletz47** [comment]
I'll add as many errors, descriptions and solutions as I can over time. It is a bit difficult as I'm not sure of a way to figure out the .msbt message each individual error code uses by utilising .rpl libraries and error.rpx. It's probably possible. I asked Maschell and he said, I quote "no idea", so I'm not sure if it's an avenue worth looking into. Putting that aside, I can continue to compile information available from what I can access, the internet and users on the Pretendo Discord server. 

I'd like to have a good chunk of error codes with at least filled out messages, descriptions and solutions in priority descending in that order. (I mean, realistically solutions would be the priority, but I've mostly listed those in order of most accomplishable descending)

You can merge it as I go if you'd like, because this may take a bit of time.

**Comment by jonbarrow** [comment]
> I'll add as many errors, descriptions and solutions as I can over time
> ...
> You can merge it as I go if you'd like, because this may take a bit of time.

In cases like this where you want to make many, many, changes, we would prefer separate PRs. You can work on them at the same time, and we can review them in parallel. You can do this fairly easily by just making different feature branches (such as `errors-160` or something) and using gits `worktree` feature to have multiple branches checked out at the same time. Then you can work on multiple PRs simultaneously without needing to keep checking them in and out.

Splitting the changes up into multiple, smaller, PRs this way helps us during review. Right now this single PR has almost ***100*** changed files, making it rather cumbersome to review all at once.

> It is a bit difficult as I'm not sure of a way to figure out the .msbt message each individual error code uses by utilising .rpl libraries and error.rpx. It's probably possible. I asked Maschell and he said, I quote "no idea", so I'm not sure if it's an avenue worth looking into.

If you're looking at actual MSBT files then https://github.com/PretendoNetwork/nintendo-file-formats/ supports decoding those. This script is what I used to generate the original base version of these JSON files.

That being said, it sounds like you may be a little confused. The MSBT files do not exist inside the RPX/RPL, they're external files. These files are generic labeled string containers used by many titles for storing strings. The error messages found here were extracted from the English MSBT files found in the `ErrEula` title (0005001B-10051000), which are found in Yaz0 compressed SARC archives inside the `content` folder. There are older languages here but they weren't used since I couldn't translate the descriptions and solutions portions.

If you're looking inside an RPX/RPL you're going to want to use Ghidra. An RPX/RPL are more or less the same thing (for these purposes), so you can treat them the same. Use https://github.com/Maschell/GhidraRPXLoader and then open the RPX/RPL (you may need to use an older version of Ghidra, unsure if Maschell's loader works for the latest release yet. It didn't last time I checked and required building from source).

RPXs/RPLs are compressed so you can also decompress them using something like https://github.com/0CBH0/wiiurpxtool. Then you can extract all the strings, though this won't help you get the error codes.

If you're using Ghidra just look around for the strings and see under what conditions that are used. For example I just opened `error.px` from the `Error (OSv255)` title (00050030-100111FF) and immediately found strings like:

```
The USB storage device has been removed,\nor is not receiving enough power.\n\nPlease press and hold the POWER Button for\nat least four seconds to turn the console off,\nand then check that the device is connected\nproperly. If it requires its own power supply,\nmake sure that is properly connected as well.
```

Which is at address 0x10000E74. This string is only referenced once at address 0x100093A8, which looks like an array entry for the global array `PTR_s_Error_Code_10009390` found at address 0x10009390. `PTR_s_Error_Code_10009390` is only ever referenced inside the function `FUN_020008ac`. This function takes in quite a lot of arguments, but it's pretty easy to see that it's taking in some details about an error, including the error code itself, and then pulling different strings to display based on the error details.

Inside `FUN_020008ac` there is an argument `param_10` which looks to be the error code itself as a uint32. This argument is used to create the variables `uVar2`, `uVar5`, and `uVar6` like so:

```cpp
uVar2 = param_10[1] / 10000;
uVar5 = param_10[1] % 10000;
uVar6 = uVar5 % 100;
```

Assuming `param_10` is a value such as 1022613 (error code 102-2613 as a uint32), then:

- `uVar2` = The error module (102)
- `uVar5` = The error code (2613)
- `uVar6` = The error codes lower value (13)

```cpp
#include <iostream>

int main() {
    uint uVar2;
    uint uVar5;
    uint uVar6;
    uint param_10 = 1022613;
    
    uVar2 = param_10 / 10000;
    uVar5 = param_10 % 10000;
    uVar6 = uVar5 % 100;

    std::cout << uVar2 << "\n"; // 102
    std::cout << uVar5 << "\n"; // 2613
    std::cout << uVar6 << "\n"; // 13

    return 0;
}
```

These variables are then used, along with `param_11`, to pull different global lists from memory and (sometimes) change the value of `uVar6`, and then either `param_11` or `uVar6` is used as an index into the arrays (depending on the array) to get the appropriate string. This string seems to then be sent to whatever shows it on screen. I'm unsure what `param_11` is right now though.

So if you want to get the error codes and messages from an RPX/RPL, you're going to have to get your hands dirty and do some reverse engineering.

**Comment by Spletz47** [comment]
I'll be sure to make new PRs for sets of error codes and add to them individually. Does that sound ok to you? Also, what I meant with the MSBT files is there obviously there is something in the code dictating what error codes use what message. It would probably require a lot of unrealistic decomplication, but it was just a thought. I'm already very aware of Ghidra and GhidraRPXLoader, which is what I used to find those more critical errors I told you about before. I have very limited knowledge with the C languages (or any language for that matter), as I'm much more experienced with RomFS. On that note, I do thank you for the explanations towards the end of your message, as it clears a lot up about that. 

Just to be clear, I was interested in a method of figuring out what messages are used for the error codes with either empty MSBT entries, or no MSBT entries at all for the respective error code (which is the reason why so many are empty in the first place).

Another little thing--this probably means nothing--but there's actually basically a copy of ErrEula in AllMessage, in SceneMsg > MsgScene. No idea why. Just worth mentioning since you mentioned ErrEula.

We can discuss this on Discord further if you'd like, but I understand you're very busy lol. Just trying to help with Bandwidth and support for users.

**Pull request closed without merging**